### PR TITLE
Allow code search by filename

### LINKED ISSUE #32096: Code search should look for the filename as well
### Feature Description

How do you guys feel about the code search feature taking the filenames into account? 

Right now, Gitea only searches for the contents of the source files. I think the search should also look for the given criteria in the filenames (this is how Bitbucket and GitHub do it).

What do you guys think about it?

### Screenshots

The screenshot below shows an excerpt of the [GitHub docs](https://docs.github.com/en/search-github/github-code-search/understanding-github-code-search-syntax) that explains how the filename also searched

image

### PR DESCRIPTION
**Pull request by bsofiato**
This is a large and complex PR, so let me explain in detail its changes.

First, I had to create new index mappings for Bleve and ElasticSerach as the current ones do not support search by filename. This requires Gitea to recreate the code search indexes (I do not know if this is a breaking change, but I feel it deserves a heads-up).

I've used [this approach](https://www.elastic.co/guide/en/elasticsearch/reference/7.17/analysis-pathhierarchy-tokenizer.html) to model the filename index. It allows us to efficiently search for both the full path and the name of a file. Bleve, however, does not support this out-of-box, so I had to code a brand new [token filter](https://blevesearch.com/docs/Token-Filters/) to generate the search terms.

I also did an overhaul in the `indexer_test.go` file. It now asserts the order of the expected results (this is important since matches based on the name of a file are more relevant than those based on its content). 
I've added new test scenarios that deal with searching by filename. They use a new repo included in the Gitea fixture. 

The screenshot below depicts how Gitea shows the search results. It shows results based on content in the same way as the current version does. In matches based on the filename, the first seven lines of the file contents are shown (BTW, this is how GitHub does it).

image

Resolves #32096

### CODE REVIEW

File: tests/gitea-repositories-meta/org42/search-by-path.git/description [hunk 1]
```diff
@@ -0,0 +1 @@
+Unnamed repository; edit this file 'description' to name the repository.
```
Comment (delvh) [inline tests/gitea-repositories-meta/org42/search-by-path.git/description]: Could you perhaps add a link here to where you got this repo from?
I really don't like not knowing how the repo looked like without major trouble.

File: modules/indexer/code/internal/util.go [hunk 2]
```diff
@@ -30,3 +34,16 @@ func FilenameOfIndexerID(indexerID string) string {
 	}
 	return indexerID[index+1:]
 }
+
+func FilenameMatchIndexPos(content string) (int, int) {
```
Comment (delvh) [inline modules/indexer/code/internal/util.go:39]: Docs!
I'm unable to guess what this method is doing from its name, return type and implementation.
Everything about this method seems weird:
I have no idea why you return the index of the `7th` newline.
It sounds like you have a clear fileformat in front of your eyes, but I don't.

Comment (delvh) [reply modules/indexer/code/internal/util.go:39]: Also, why is the first return value always `0`?

File: modules/indexer/internal/bleve/util_test.go [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+			Input:     "Geschwindigkeit",
+			Fuzziness: 2,
+		},
+		{
+			Input:     "non-exist",
+			Fuzziness: 0,
+		},
+		{
+			Input:     "갃갃갃",
+			Fuzziness: 0,
```
Comment (delvh) [inline modules/indexer/internal/bleve/util_test.go:36]: I have no idea what these numbers mean.
Perhaps we should use constants instead?

File: modules/indexer/code/bleve/token/path/path.go [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+
+type TokenFilter struct{}
+
+func NewTokenFilter() *TokenFilter {
+	return &TokenFilter{}
+}
+
+func TokenFilterConstructor(config map[string]any, cache *registry.Cache) (analysis.TokenFilter, error) {
+	return NewTokenFilter(), nil
+}
```
Comment (delvh) [inline modules/indexer/code/bleve/token/path/path.go:26]: This feels… quite Java-y.
Why?
And why these useless return values?

File: modules/indexer/code/bleve/token/path/path.go [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+		return generatePathTokens(input, false)
+	}
+
+	normal := generatePathTokens(input, false)
+	reversed := generatePathTokens(input, true)
+
+	return append(normal, reversed...)
+}
+
+func generatePathTokens(input analysis.TokenStream, reversed bool) analysis.TokenStream {
```
Comment (delvh) [inline modules/indexer/code/bleve/token/path/path.go:46]: Docs!

File: modules/indexer/code/bleve/token/path/path.go (hunk 4 above)
Comment (delvh) [reply modules/indexer/code/bleve/token/path/path.go:26]: Why not just initialize a `&TokenFilter{}` directly instead?

File: modules/indexer/code/internal/util.go (hunk 2 above)
Comment (bsofiato) [reply modules/indexer/code/internal/util.go:39]: Whenever there is a filename-based fetch, it will return the first seven lines of the file (BTW, this is mimics the github behavior).  This is used by the page to render the code snippet associated with the search result. 

So the start value will be always be zero because it the code snippted always start at the begining of the file when the hit is based on filename.

File: modules/indexer/code/bleve/token/path/path.go (hunk 4 above)
Comment (bsofiato) [reply modules/indexer/code/bleve/token/path/path.go:26]: While I agree its kind weird, I decided to just follow the same pattern bleve uses on its token filters (as shown [here](https://github.com/blevesearch/bleve/blob/master/analysis/token/ngram/ngram.go), [here](https://github.com/blevesearch/bleve/blob/master/analysis/token/reverse/reverse.go), and [here](https://github.com/blevesearch/bleve/blob/master/analysis/token/lowercase/lowercase.go)). 

If you guys think it is better to change it, no problem :)

File: tests/gitea-repositories-meta/org42/search-by-path.git/description (hunk 1 above)
Comment (bsofiato) [reply tests/gitea-repositories-meta/org42/search-by-path.git/description]: I've just copied over a fixture repository (BTW copied the repo which was used by the current scenarios) and added a bunch of more files. 

Can you provide some examples of what kind of information do you think would be nice to have on this file ? I've checked the other repos on tests/gitea-repositories-meta and did not found only two repositories that provided a description that was not `Unnamed repository; edit this file 'description' to name the repository.` (in that case, it was `The repository will be used to test third-party renderer in TestExternalMarkupRenderer`). 

Do you think that `This repository will be used to test code search` suffice ?

File: modules/indexer/internal/bleve/util_test.go (hunk 3 above)
Comment (bsofiato) [reply modules/indexer/internal/bleve/util_test.go:36]: These are the expected fuzzyness for each string. However, I'm not quite sure on how to name the constant :( (if you guys have any ideias, I'm open)

A little more context, the method `GuessFuzzinessByKeyword` already [exists on gitea codebase](https://github.com/go-gitea/gitea/blob/main/modules/indexer/internal/bleve/util.go#L51) but lacks a unit tests. As of today, it calculates the fuzziness based on the size of the whole search criterion (unless some non-standard characters are used, in such a case, it defaults to zero). 

What I did was simply calculate the fuziness based on the terms instead of the whole string. This prevent that a criterion composed of a bunch of small words to have high degree of fuzziness. A fuzziness degree was rendering bleve too leninent and making it return results that would make much sense (e.g., [this](https://github.com/go-gitea/gitea/blob/05a5090c34530d0bcfd62776cbfd42f6da3386bf/modules/indexer/code/indexer_test.go#L80) scenario would return some results when it should return none)

File: modules/indexer/code/bleve/token/path/path.go (hunk 5 above)
Comment (bsofiato) [reply modules/indexer/code/bleve/token/path/path.go:46]: Docs added

File: modules/indexer/code/internal/util.go (hunk 2 above)
Comment (bsofiato) [reply modules/indexer/code/internal/util.go:39]: Included some docs as well

**Comment by yp05327** [comment]
The year of copyright in new files should be 2024.

**Comment by bsofiato** [comment]
> The year of copyright in new files should be 2024.

Done

**Comment by lafriks** [comment]
Please do not force push, it's hard to follow changes this way

**Comment by bsofiato** [comment]
> Please do not force push, it's hard to follow changes this way

Sure. My bad 😔

File: tests/gitea-repositories-meta/org42/search-by-path.git/description (hunk 1 above)
Comment (delvh) [reply tests/gitea-repositories-meta/org42/search-by-path.git/description]: Yes, but that's exactly the problem I can see:
In my experience so far, most of these test repos have been cloned from try.gitea.io or gitea.com at some point.
However, since we don't know from where exactly, we can only guess how the original repo looked like without going through the trouble of converting a non-`.git` folder into one.
Thus, for the vast majority I have no clue what files/commit history they even offer…
But sure, if you don't know either, then it is not needed.

File: modules/indexer/internal/bleve/util_test.go (hunk 3 above)
Comment (delvh) [reply modules/indexer/internal/bleve/util_test.go:36]: Can you perhaps add that as docs as well to the method?

File: tests/gitea-repositories-meta/org42/search-by-path.git/description (hunk 1 above)
Comment (bsofiato) [reply tests/gitea-repositories-meta/org42/search-by-path.git/description]: @delvh I see your point, and I think it is fair (as a matter of fact, I had the same issue when trying to see the contents of the repositories used by the current test scenarios). What do you think about putting the directory structure of the given repo in its description? I was thinking of something like the snippet below:

```
my-app/
├─ node_modules/
├─ public/
│  ├─ favicon.ico
│  ├─ index.html
│  ├─ robots.txt
├─ src/
│  ├─ index.css
│  ├─ index.js
├─ .gitignore
├─ package.json
├─ README.md
```

File: modules/indexer/internal/bleve/util_test.go (hunk 3 above)
Comment (bsofiato) [reply modules/indexer/internal/bleve/util_test.go:36]: Done. Added some docs to both the test and the `GuessFuzzinessByKeyword` method.

File: tests/gitea-repositories-meta/org42/search-by-path.git/description (hunk 1 above)
Comment (delvh) [reply tests/gitea-repositories-meta/org42/search-by-path.git/description]: Sounds good

Comment (bsofiato) [reply tests/gitea-repositories-meta/org42/search-by-path.git/description]: @delvh I've added the directory structure to the description file :)

File: modules/indexer/code/elasticsearch/elasticsearch.go [hunk 6]
```diff
@@ -244,35 +283,36 @@ func indexPos(content, start, end string) (int, int) {
...
+		}
+
 		// FIXME: There is no way to get the position the keyword on the content currently on the same request.
 		// So we get it from content, this may made the query slower. See
 		// https://discuss.elastic.co/t/fetching-position-of-keyword-in-matched-document/94291
 		var startIndex, endIndex int
-		c, ok := hit.Highlight["content"]
-		if ok && len(c) > 0 {
+		if c, ok := hit.Highlight["filename"]; ok && len(c) > 0 {
+			startIndex, endIndex = internal.FilenameMatchIndexPos(res["content"].(string))
```
Comment (lunny) [inline modules/indexer/code/elasticsearch/elasticsearch.go:303]:
```suggestion
			startIndex, endIndex = internal.FilenameMatchIndexPos(res["filename"].(string))
```

Comment (bsofiato) [reply modules/indexer/code/elasticsearch/elasticsearch.go:303]: This is indeed `res["content"]`.

The `hit.Highlight["filename"]` statement returns `true` when the given result is related to a match in the filename. In such a case, we render the first 7th lines of the file as the result. The `FilenameMatchIndexPos` function computes the boundaries of such a snippet using. For this, it needs the file content -- `res["content"]`.

**Review by delvh — APPROVED** [review]

**Comment by delvh** [comment]
>  This requires Gitea to recreate the code search indexes

Ah, right - for this to happen, we need to increase `issueIndexerLatestVersion` in the affected indexers

**Comment by lunny** [comment]
Can you add a search test with filename as keyword?

**Comment by bsofiato** [comment]
> Can you add a search test with filename as keyword?

I have included some., I'll provide some comments on the test cases to better describe their underlying scenarios. What do you think ?

**Comment by bsofiato** [comment]
> > Can you add a search test with filename as keyword?
> 
> I have included some., I'll provide some comments on the test cases to better describe their underlying scenarios. What do you think ?

Hey @lunny, I've added some comments to the test cases to better describe their scenarios :)

**Review by lunny — APPROVED** [review]
